### DISCUSSION TITLE
How I Built a Serverless Search for My Blog - Gunnar Morling

### DISCUSSION THREAD

**Discussion by blog-comment-importer** [gunnarmorling/discussions.morling.dev#276] [Announcements]
https://www.morling.dev/blog/how-i-built-a-serverless-search-for-my-blog/

**Comment by blog-comment-importer** [comment]
_**Will Clark** wrote at 2020-08-06T03:16:01Z (comment imported from Disqus):_

<p>Gunnar, I just wanted to say thank you for writing this up. I've heard the FaaS buzz for a while now... this blog really helped me "get it" with a practical use case. I'm gonna to give it a go sometime.</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2020-08-06T12:09:51Z (comment imported from Disqus):_

<p>Thanks a lot, Will, for taking your time to read the post and providing that feedback. Very glad to hear it's useful for you, appreciating the nice words!</p>

**Comment by blog-comment-importer** [comment]
_**Joost van der Griend** wrote at 2020-08-07T13:15:12Z (comment imported from Disqus):_

<p>Thank you for this article! It is really good that you add your reasoning in here. I have similar convictions (free, adds, managing a VM, etc), and this gives me inspiration on doing something similar.</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2020-08-07T14:12:26Z (comment imported from Disqus):_

<p>Thanks for the great feedback, Joost! Good luck for your Serverless adventures.</p>

**Comment by blog-comment-importer** [comment]
_**StuartWinchester** wrote at 2020-09-11T11:00:04Z (comment imported from Disqus):_

<p>Great article Gunnar!  I'd love to know your thoughts on how something similar could be setup in a pure Openshift/Kubernetes environment with Knative.  Is there anything that can act as the Amazon API Gateway and trigger Knative stuff?</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2020-09-11T19:01:53Z (comment imported from Disqus):_

<p>Hey Stuart, thanks for the nice feedback! Great question re API Gateway; it depends a bit on what your requirement is. It's not needed per-se with Knative, which will take care of the request routing, spinning up the instances etc. You might consider adding an API gateway such as 3Scale in front though for things like rate limiting, usage plans, API keys etc.</p>

**Comment by blog-comment-importer** [comment]
_**Octavian** wrote at 2021-02-13T23:07:20Z (comment imported from Disqus):_

<p>Such a lovely write-up... so many things revealed! Thanks!</p>

**Reply by blog-comment-importer** [reply]
_**Octavian** wrote at 2021-02-14T07:28:31Z (comment imported from Disqus):_

<p>Although you might want, indeed, to tune a bit Lucene. If I search for "Elasticsearch", I get correct  results but if I search for "elastic", I don't get anything back ;)</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2021-02-14T09:13:40Z (comment imported from Disqus):_

<p>Hey thanks for the nice words! Great to hear the post is useful to you.</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2021-02-14T09:16:38Z (comment imported from Disqus):_

<p>Ah, that's interesting. Might need the right option to find subword matches. Thanks for pointing this out!</p>

**Comment by blog-comment-importer** [comment]
_**Luke Thompson** wrote at 2021-06-01T01:17:24Z (comment imported from Disqus):_

Awesome article!</p><p>FYI as another data point on performance, here's what it looks like from Australia :) 🇦🇺:

![disqus_comment](https://user-images.githubusercontent.com/28612/141851506-caae50eb-e1e0-4c80-9970-00d286c83eda.png)

<p>You do feel the search response time a little, being almost a second. Even so it's pretty cool. I love the fact you've built this!</p><p>I had a couple of thoughts about CORS:<br>1) It would be good if you could cache the CORS request, so that the ping established permissions to the search domain only once, and this didn't need to be repeated for every API call.<br>Ref: <a href="https://httptoolkit.tech/blog/cache-your-cors/" rel="nofollow noopener" title="https://httptoolkit.tech/blog/cache-your-cors/">https://httptoolkit.tech/bl...</a></p><p>2) Since you have CloudFlare fronting both the blog and the API, why not just have the search API presented by CloudFlare on the same domain on a path like /api/search on <a href="http://morling.dev?" rel="nofollow noopener" title="morling.dev?">morling.dev?</a> This way you can avoid all the preflight OPTIONS requests, and also remove the concern you had about additional costs on API Gateway. So long as you configure a custom domain for the API of <a href="http://morling.dev" rel="nofollow noopener" title="morling.dev">morling.dev</a> and ensure the path prefix is the same it should work well.</p><p>On another topic, since it is serverless and you only pay for consumption, why not make the deployment global in a couple of other regions so that everyone in the world gets &lt;100ms search response times? It shouldn't cost any more I think and only adds a little complexity to the deployment. If you're going to over-engineer a custom search engine for your blog why not do the thing properly :). <br>Ref: <a href="https://aws.amazon.com/blogs/compute/building-a-multi-region-serverless-application-with-amazon-api-gateway-and-aws-lambda/" rel="nofollow noopener" title="https://aws.amazon.com/blogs/compute/building-a-multi-region-serverless-application-with-amazon-api-gateway-and-aws-lambda/">https://aws.amazon.com/blog...</a></p><p>I think one of the things I find really compelling about the serverless model is this capacity to deploy lots of versions / regions of your software, make it very efficient and scale globally + have a very low operating cost.</p><p>Thanks very much for taking such a thorough and deep dive into building a native serverless solution with Java. I learnt a lot! <br>(I'm also very much looking forward to <a href="https://mail.openjdk.java.net/pipermail/discuss/2020-April/005429.html" rel="nofollow noopener" title="https://mail.openjdk.java.net/pipermail/discuss/2020-April/005429.html">Project Leyden</a> too)</p>

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2021-06-01T21:04:42Z (comment imported from Disqus):_

<p>Hey Luke, thanks a lot for your nice comment and the great input. This sort of feedback and interaction is really motivating!</p><p>Thx for sharing the latency to Australia, totally can see that it would be that long from where this is hosted.</p><p></p><blockquote>It would be good if you could cache the CORS request...</blockquote><p></p><p>Ah, that's a great point, i didn't think of this yet. I should take a look.</p><p></p><blockquote>Since you have CloudFlare fronting both the blog and the API...</blockquote><p></p><p>I'm using CloudFlare only for the API, not the blog itself. The latter is deployed to GitHub Pages with a custom domain and so far I hadn't seen the need for putting the blog to CF, too. As you say, it should avoid all the CORS hassle, so it may be worth considering actually.</p><p></p><blockquote>why not make the deployment global in a couple of other regions so that everyone in the world gets &lt;100ms search response times?<br></blockquote><p></p><p>Yeah, I had been thinking about this, too, but so far I couldn't motivate myself to tackle this task. Lambda @ Edge (is that's how it's called?) could be an option, though I always felt it's really hard to keep track of resources deployed to multiple AWS regions.</p><p>An alternative option I'm contemplating over is to move off of Lambda to alternatives like <a href="http://fly.io" rel="nofollow noopener" title="fly.io">fly.io</a>, which should make it really simple to deploy this sort of read-only workload to multiple regions. Thanks to the way the Quarkus build, that'd require only to switch build profiles for producing a standard container image instead of the Lambda ZIP.</p><p></p><blockquote>I think one of the things I find really compelling about the serverless model is this capacity to deploy lots of versions / regions of your software</blockquote><p></p><p>Absolutely; I think there's a great future to this sort of architecture. I like the way of <a href="http://fly.io" rel="nofollow noopener" title="fly.io">fly.io</a> of describing this as "CDN for applications". The elephant in the room is mutable state of course. If you need to synchronize on one canonical data store, write latencies can become a bottleneck (reads can be served well from local caches). Luckily, that's not a problem for my use case here :)</p>

**Comment by fafriat** [comment]
Hi and thanks for this very interesting post.

You describe the workflow to configure Lucene for native image compilation...

There is an alternative way you may have used to automatically generate the configuration by using the [GraalVM Tracing Agent](https://www.graalvm.org/reference-manual/native-image/Agent/) but this may take some time to configure with Quarkus...

This is one of the reason I created the micro-framework [Microlam](https://microlam.io), where running your project with this java agent is included in the development workflow and available directly from maven.

Another point, that could reduce the costs of your service, would have been to deploy to ARM, this is also available from maven using Microlam adding -Dnative=java11-arm64 or java17-arm64 will enable this.

I would be happy if you could give Microlam a try and give me your feedback...

**Comment by iainelder** [comment]
I found this by searching Google for `"arn:aws:cloudformation:eu-central-1:aws:transform/Serverless-2016-10-31"`. Yours is the first of just two results!

The `transform` resource type is not documented in the [CloudFormation service authorization reference](https://docs.aws.amazon.com/service-authorization/latest/reference/list_awscloudformation.html).

And yet I see errors in my deployments similar to `User: ... is not authorized to perform: cloudformation:CreateChangeSet on resource: arn:aws:cloudformation:eu-central-1:aws:transform/Serverless-2016-10-31 because no identity-based policy allows the cloudformation:CreateChangeSet action`

Are you aware of any other documentation on the resource?

For now, your sample policy seems to be the best documentation known to Google.
